lib/vdsm/storage/mpathconf.py

```diff
@@ -428,3 +432,95 @@ def read_metadata():
         return Metadata(REVISION_OLD, private)
 
     return Metadata(REVISION_MISSING, private)
+
+
+def check_local_config():
+    """
+    Check if 'user_friendly_names' is enabled in the multipathd configuration.
+    The configuration is obtained by running the multipathd command to show
+    the local configuration and capturing the output.
+    Parse the output and check that the contents are valid.
+
+    Raises:
+        exception.InvalidConfiguration: if any section have
+            'user_friendly_names' enabled. Contains the offending section name.
+    """
+    mpathconf = []
+    _parse_conf(multipath.show_config_local().split('\n'), mpathconf)
+    for conf in mpathconf:
+        failed_section, check_ok = _check_conf_attrs(conf.name, conf.attrs)
+        if not check_ok:
+            if failed_section is None:
+                failed_section = conf
+            raise exception.InvalidConfiguration(
+                reason=("'user_friendly_names' is enabled "
+                        f"in multipath configuration. Check '{CONF_FILE}'."),
+                section=failed_section)
+        # Keep looping to ensure 'user_friendly_names' is not enabled by
+        # any other device.
+
+
+def _check_conf_attrs(name, attributes):
+    """
+    Check that 'user_friendly_names' is disabled in the attributes section
+    of the parsed multipathd configuration.
+    Search is recursive.
+
+    Args:
+        name (str): name of the section to print.
+        attributes: attributes in the configuration grammar can contain
+            either nested attributes or key-value paris.
+
+    Returns:
+        Tuple[ConfSection, bool]:
+            ConfSection: used if any nested section have 'user_friendly_names'
+                enabled. Contains the offending section tuple.
+            bool: all 'user_friendly_names' are disabled
+    """
+    for attr in attributes:
+        if isinstance(attr, ConfPair):
+            if attr.key == 'user_friendly_names' and attr.value == 'yes':
+                return None, False
+        else:
+            assert isinstance(attr, ConfSection)
+            # Nested attributes (recursive call)
+            _, check_ok = _check_conf_attrs(
+                f"{name}.{attr.name}", attr.attrs)
+            if not check_ok:
+                return attr, check_ok
+    return None, True
+
+
+def _parse_conf(mpath_out, parsed_conf, depth=0):
```

Comment:
We can simplify this by accepting an iterator instead of list of lines.

For example the caller can use:

     import io
     it = io.StringIO(out)
     tree = _parse_conf(it)

The advantage of the iterator is that we don't need to manage the depth or pass the list to append value into. Instead we can just return the parsed value.

Reply:
Nice, I am not very familiar with the io functions, so I didn't know this could be done so easily.

This will make parsing a lot easier, thanks!